activerecord/lib/active_record/associations/builder/singular_association.rb

```diff
@@ -19,6 +19,10 @@ def self.define_accessors(model, reflection)
         def reload_#{name}
           association(:#{name}).force_reload_reader
         end
+
+        def #{name}!
+          raise ActiveRecord::RecordNotFound unless association(:#{name}).force_reload_reader.present?
```

Comment:
```suggestion
          raise ActiveRecord::RecordNotFound unless association(:#{name})
```

I don't know why we are reloading, but I don't think we should since `model.foo = bar; model.foo!" should not raise, and if we reload it will.

We should add a test for it.

Reply:
@rafaelfranca when I removed the call to reload, my [original test](https://github.com/rails/rails/pull/47136/files#diff-2a3debce0428327feafac8cf69f6813284718e3e5ad7617c62d28e7baeea91a8R40-R46) failed. I think this is because my test is deleting a record in the set-up step.

Let me know if I'm understanding this correctly.